Builtins expose Enso methods

Akirathan opened the pull request:
Closes #11589

### Pull Request Description

Builtin types expose methods via the interop protocol. In fact, all types expose methods via interop protocol.

To get and invoke a method on a builtin object via interop protocol:
```java
var meta = interop.getMetaObject(text);
meta.invokeMember(meta, "+", ...);
```
Example in [InvokeBuiltinMethodViaInteropTest](https://github.com/enso-org/enso/blob/def1f6b7247bebf2060f1a833c6b9cf43d24b868/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/InvokeBuiltinMethodViaInteropTest.java#L129-L150)

To get and invoke a method on a builtin object via `org.graalvm.polyglot.Value` (this can be done, e.g., in std libs in Java code):
```java
Value object;
Value meta = object.getMetaObject();
meta.invokeMember(...);
```
Example in [BuiltinTypesExposeMethodsTest](https://github.com/enso-org/enso/blob/752ac901a54757b329cbdc229ee2795bdc9a12e7/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/BuiltinTypesExposeMethodsTest.java#L76-L105).

### Important Notes
Method members are reported in [Type.getMembers](https://github.com/enso-org/enso/blob/3969177ef23276d111b06a79f749dd2c69e86869/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java#L350-L370) as **internal** members.
This has a side effect that in the debugger, types have method properties (but not atoms):
![image](https://github.com/user-attachments/assets/ce12648c-04f5-4af6-af6b-c2c87826c59f)


The first attempt to implement `hasMember` and `getMembers` on `BuiltinObject` abandoned - https://github.com/enso-org/enso/pull/11687#issuecomment-2539743567

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.
- [x] If meaningful changes were made to logic or tests affecting Enso Cloud integration in the libraries, 
      or the Snowflake database integration, a run of the [Extra Tests](https://github.com/enso-org/enso/actions/workflows/extra-nightly-tests.yml) has been scheduled.
  - If applicable, it is suggested to paste a link to a successful run of the Extra Tests.

Commit:
Add BuiltinsExposeMethodsTest

Commit:
typo in docs

Commit:
Test iteartes all the builtins and checks that all methods are invocable member

Commit:
Add TypesExposeConstructorsTest

Commit:
Merge branch 'develop' into wip/akirathan/11589-builtins-expose-methods

Commit:
Naive implementation of BuiltinObject base class.

Commit:
Ref extends BuiltinObject

Commit:
Builtin type anot processor ensures that a class must extend BuiltinObject

Commit:
Enrich BuiltinsExposeMethodsTest

Commit:
All builtin types extend BuiltinObject

Commit:
Text is BuiltinObject

Commit:
EnsoBigInteger is BuiltinObject

Commit:
BuiltinObject does no asserts in constructor

Commit:
ArrayProxy is BuiltinObject

Commit:
Test skips host values and Nothing

Commit:
fmt

Commit:
Remove outdated test.

No longer true what this test tested

Commit:
EqualsComplexNode: Timezone and duration are not treated as object with members

Commit:
Fix DebuggingEnsoTest - Date in js is date time, not date

Commit:
Fix DateTest - ensoDate now has members

Commit:
Add interop.readMember test to BuiltinsExposeMethodsTest

Commit:
VectorSortTest is executed in context

Commit:
Add tests that invoke builtin methods on particular builtin types

Commit:
member methods on BuiltinObject are behind TruffleBoundary

Akirathan commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsComplexNode.java:
```diff
@@ -543,6 +543,12 @@ boolean isObjectWithMembers(Object object, InteropLibrary interop) {
     if (interop.isTime(object)) {
       return false;
     }
+    if (interop.isDuration(object)) {
```
Both TimeZone and Duration now have members, so we need to add these guards here.

JaroslavTulach replied:
The whole `EqualsXyzNode` set of `@Specialize` annotation is fragile, unreadable, not fully testable and deserves a rewrite. At least that's the feeling I get every time I am changing the code. 

Such a gentle change is OK, unless total rewrite is attempted.

Akirathan replied:
See my answer in https://github.com/enso-org/enso/pull/11687#issuecomment-2528845368 - yes it is a mess.

Commit:
Cache builtin type in BuiltinObject.

This fixes native image build of engine-runner.

Commit:
Reuse context in DebuggingEnsoTest.

This fixes the AssertionError in invalid sharing layer.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -15,40 +13,20 @@
-@ExportLibrary(TypesLibrary.class)
 @Builtin(pkg = "date", name = "Duration", stdlibName = "Standard.Base.Data.Time.Duration.Duration")
-public final class EnsoDuration extends EnsoObject {
+public final class EnsoDuration extends BuiltinObject {
```
Introducing a shared _super class_ to all builtins is a good idea. It will help us enforce common behavior when needed.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -15,40 +13,20 @@
   private final Duration duration;
 
   public EnsoDuration(Duration duration) {
+    super("Duration");
```
There already is `@Builtin` annotation at line 21 and it specifies full `stdLibName`. Can't we re-use it somehow?

Akirathan replied:
Done in 8acb04a4f948ab28ea028c55e6035ea35256799e

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinObject.java:
```diff
@@ -0,0 +1,156 @@
+package org.enso.interpreter.node.expression.builtin;
```
New package? Why `node.expression`? This class is `TruffleObject`, it is neither _expression_, neither _node_.

I have a long time desire to split `runtime` into two projects:
- `runtime`
- `runtime-builtins`

I have already tried it few times and the split is not easy. It is all too convoluted. Last time I at least reported:
- #11593

We cannot (and shouldn't) try to break the module to two parts in this PR, but it'd be good to keep in mind we want to do it. E.g. **collocate builtin functionality** into some _sub-package_.

Akirathan replied:
Moved to package `org.enso.interpreter.runtime.builtin` in f187211da9f44db706220e61d2442013f581e0c7.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinObject.java:
```diff
@@ -0,0 +1,156 @@
+
+  /**
+   * @param builtinName Simple name of the builtin that should be contained in {@link
+   *     org.enso.interpreter.runtime.builtin.Builtins#builtinsByName}.
```
I am dreaming of creating a _prelude_:
- #8852

A _"prelude" library_ one can use without importing whole (bloated) `Standard.Base` and still be productive with Enso builtins. I am not really sure how to do it, but:

- I feel we should associate each builtin with its `stdLibName`
- (maybe) generate prelude sources for such builtins automatically

Thus I would prefer full `stdLibName` here. But I don't have a solid justification...

Akirathan replied:
Out of scope for this PR, as it would require changes to how [Builtins](https://github.com/enso-org/enso/blob/d687365070381b07eb4214bc665a13912c3bbb96/engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Builtins.java) is initialized. But I share your gut feeling.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinObject.java:
```diff
@@ -0,0 +1,156 @@
+  }
+
+  @ExportMessage
+  public boolean hasMembers() {
```
Do we need these methods to be `public`? If not, make them _package private_. If they have to be `public`, can we make them `final`? I don't want calls to `super.getMembers` & other `super.calls` spread around our code base. Such calls usually leads to [object-oriented spaghetti mess](http://wiki.apidesign.org/wiki/NetBeans_Runtime_Container#Modularity_Manifesto).

Akirathan replied:
They cannot be made package private:
```
[info] [FrgaalJavaCompiler] compiling 433 Java sources with module-info.java to /home/pavel/dev/enso/engine/runtime/target/scala-2.13/classes ...
[error] /home/pavel/dev/enso/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDate.java:20:1:  error: Found invisible exported elements in super type 'BuiltinObject': 
[error] public final class EnsoDate extends BuiltinObject {
[error]              ^     - org.enso.interpreter.runtime.builtin.BuiltinObject.hasMembers()
[error]   Increase their visibility to resolve this problem.
```

But I at least made them `final` in 6f31ebfff7e2ec6f88d9b3a374150484c8253567

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinObject.java:
```diff
@@ -0,0 +1,156 @@
+  }
+
+  @ExportMessage
+  @TruffleBoundary
```
`@TruffleBoundary` for `readMember`? That's not going to be fast and we want `readMember` to be fast.

Akirathan replied:
Optimized in e441faba5b21b4d38498edc3a0bb614a35c5b64b

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinObject.java:
```diff
@@ -0,0 +1,156 @@
+
+  @ExportMessage
+  @TruffleBoundary
+  public Object invokeMember(String member, Object[] args)
```
Can we remove `@TruffleBoundary` from `invokeMember`?

Akirathan replied:
Removed boundaries from all the methods in e441faba5b21b4d38498edc3a0bb614a35c5b64b

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/BuiltinObject.java:
```diff
@@ -0,0 +1,156 @@
+  }
+
+  private Set<String> methodNames() {
+    var methodNames =
```
This should be cached to make the method fast. A case for [ClassValue](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/ClassValue.html), maybe?

Akirathan replied:
Optimized in e441faba5b21b4d38498edc3a0bb614a35c5b64b

JaroslavTulach reviewed (commented):
Nice. Few comments added inline to get a discussion started.

Akirathan commented:
> The whole EqualsXyzNode set of @Specialize annotation is fragile, unreadable, not fully testable and deserves a rewrite. At least that's the feeling I get every time I am changing the code.
Such a gentle change is OK, unless total rewrite is attempted.


@JaroslavTulach I have found out the [modifications needed](https://github.com/enso-org/enso/pull/11687/commits/475819aab1b5522a62c344958afcfa859ae5a90a) in EqualsComplexNode based on the test failure in https://github.com/enso-org/enso/actions/runs/12165411858/job/33929304252#step:7:5800 which says that there is a different result from a cached version of EqualsNode vs its uncached version. Without that test, we would experience very nasty failures later on. So yes, I agree that it is a mess that deserves a rewrite.

<blockquote><img src="https://opengraph.githubassets.com/8a299651669d7391f9f7c1224996cb46209b3aa27a307fc65a0281becd91c732/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Builtins expose Enso methods · enso-org/enso@5a8ef4e</a></strong></div><div>Enso Analytics is a self-service data prep and analysis platform designed for data teams. - Builtins expose Enso methods · enso-org/enso@5a8ef4e</div></blockquote>

Commit:
Merge branch 'develop' into wip/akirathan/11589-builtins-expose-methods

Commit:
Reuse hardcoded builtinName constants from annotation

Commit:
Move BuiltinObject to package org.enso.interpreter.runtime.builtin

Commit:
Fix FQN of BuiltinObject in the annotation processor

Commit:
Make exported messages on BuiltinObject final

Commit:
Update generated class names in Builtins.

Fixes compilation after 8acb04a4f948ab28ea028c55e6035ea35256799e

Commit:
Exported messages in BuiltinObject are not behind TruffleBoundary

JaroslavTulach reviewed (approved):
- OK, let's integrate.

> - cannot call Ref.get: https://github.com/enso-org/enso/pull/11577/files#r1848224116
> - cannot call File.path on [EnsoFile](https://github.com/enso-org/enso/pull/11530/files#r1838708041)
> - cannot call to_text on Vector - discovered while working on
https://github.com/enso-org/enso/issues/11569

- the #11589 issue started with three examples of missing builtin methods
- shall we not fix some of these cases in this PR?

JaroslavTulach commented:
There is a lot of `Table_Tests` failures like [this one](https://github.com/enso-org/enso/actions/runs/12257340893/job/34194878025?pr=11687#step:7:2460):
```
org.enso.table.error.UnexpectedTypeException: Unexpected value type. Expected a java.time.LocalTime but got com.oracle.truffle.polyglot.PolyglotMap.
```

JaroslavTulach commented:
There is a lot of [invalid sharing layer](https://github.com/enso-org/enso/actions/runs/12257340893/job/34194876828?pr=11687#step:7:6815):
```
 Test org.enso.compiler.test.ExecCompilerTest.chainedSyntax failed: org.graalvm.polyglot.PolyglotException: java.lang.AssertionError: Invalid sharing of AST nodes detected. The current context uses a different sharing layer than the executed node. A common cause of this are CallTargets that are reused across different contexts in an invalid way.Stack trace: 
```
I'll try to reproduce the failures locally. When running `runtime-integration-tests/test` I am getting:
```
[info] Test run org.enso.interpreter.test.interop.TypesExposeConstructorsTest finished: 0 failed, 0 ignored, 8 total, 0.005s
[error] Test org.enso.interpreter.test.hash.HashCodeTest failed: java.lang.AssertionError: It is a time like value (Error: (Compile_Error.Error 'No polyglot symbol for org.enso.base.time.EnsoDateTimeFormatter')), took 0.0 sec
[error]     at org.enso.interpreter.test.ValuesGenerator.timesAndDates(ValuesGenerator.java:531)
[error]     at org.enso.interpreter.test.hash.HashCodeTest.fetchAllUnwrappedValues(HashCodeTest.java:80)
[error]     at org.enso.interpreter.test.hash.HashCodeTest.initContextAndData(HashCodeTest.java:52)
[error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]     at java.lang.reflect.Method.invoke(Method.java:580)
```
and then also the _sharing layer_ error latter. I'd start by fixing the first error and hoping the subsequent ones will go away.

Commit:
[WIP] Add BuiltinsJavaInteropTest

Commit:
Storage.vectorizedOrFallbackBinaryMap accepts Value as parameter

Commit:
Builtin types expose methods, not BuiltinObject

Commit:
BuiltinObject has no members

Commit:
Remove test Text.is_empty

Text is a builtin type with Builtins module scope, which does not have `is_empty` method.

Commit:
Fix tests - invocation is done via static methods

Commit:
Type.InvokeMember uses InvokeFunctionNode instead of UnresolvedSymbol

Commit:
fmt

Commit:
Type.InvokeMember prepends receiver argument

Commit:
Test invocation of File.path

Commit:
Test invocation of Vector.to_text

Akirathan commented:
Trying to first integrate #11835 seems like a dead end. Let's rather try a different approach - builtin objects do not expose members, but their types do. 752ac901a54757b329cbdc229ee2795bdc9a12e7 starts to implement `hasMember` and `getMembers` on a `Type` so that it returns methods as invocable members.

Akirathan commented:
> the
https://github.com/enso-org/enso/issues/11589 issue started with three examples of missing builtin methods
shall we not fix some of these cases in this PR?

@JaroslavTulach  `Ref.get`, `Vector.to_text`, and `File.path` mentioned in https://github.com/enso-org/enso/issues/11589#issue-2671967581 tested in [InvokeBuiltinMethodViaInteropTest](https://github.com/enso-org/enso/blob/def1f6b7247bebf2060f1a833c6b9cf43d24b868/engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/InvokeBuiltinMethodViaInteropTest.java)

Commit:
Method fetching is done behind truffle boundary

JaroslavTulach committed:
Reassigning @CompilationFinal field should deoptimize

Akirathan commented:
`BuiltinObject` super class split into a separate PR - #11861

JaroslavTulach commented:
@Akirathan can you merge in the most recent `develop` branch? Then we can check and see what can be done with the rest...

Commit:
Merge branch 'develop' into wip/akirathan/11589-builtins-expose-methods

# Conflicts:
#	engine/runtime-integration-tests/src/test/java/org/enso/example/PolyglotTestClass.java
#	engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/BuiltinsJavaInteropTest.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/BuiltinObject.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDate.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoFile.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoTimeOfDay.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoTimeZone.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ManagedResource.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Ref.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/EnsoHashMap.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Array.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayLikeHelpers.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/Vector.java
#	engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/Warning.java

Commit:
Fixes after merge

Commit:
Add TruffleBoundary

Commit:
Revert DateTest

Commit:
Fix PrivateConstructor test

Commit:
Methods on Type are internal members

Commit:
fmt

Commit:
Add TruffleBoundary

Commit:
Revert Storage to develop

Commit:
Remove FIXME comment

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/InvokeBuiltinMethodViaInteropTest.java:
```diff
@@ -0,0 +1,151 @@
+        ctx,
+        () -> {
+          var interop = InteropLibrary.getUncached();
+          var refUnwrapped = ContextUtils.unwrapValue(ctx, ref);
```
OK, so this is result of `Ref.new 42`.

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/InvokeBuiltinMethodViaInteropTest.java:
```diff
@@ -0,0 +1,151 @@
+          assertThat(
+              "Ref builtin object should not have any members",
+              interop.hasMembers(refUnwrapped),
+              is(false));
```
The `Ref` object has no members.

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/InvokeBuiltinMethodViaInteropTest.java:
```diff
@@ -0,0 +1,151 @@
+              "Ref should have a meta-object (Ref type)",
+              interop.hasMetaObject(refUnwrapped),
+              is(true));
+          var refMeta = interop.getMetaObject(refUnwrapped);
```
One has to ask for `Ref` object's **meta-object** - e.g. `Type`

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/InvokeBuiltinMethodViaInteropTest.java:
```diff
@@ -0,0 +1,151 @@
+          assertThat(
+              "Ref meta-object should have a 'get' method",
+              interop.isMemberInvocable(refMeta, "get"),
+              is(true));
```
and there is invocable method `get` on the `Ref` **type**.

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/builtins/InvokeBuiltinMethodViaInteropTest.java:
```diff
@@ -0,0 +1,151 @@
+              "Ref meta-object should have a 'get' method",
+              interop.isMemberInvocable(refMeta, "get"),
+              is(true));
+          var res = interop.invokeMember(refMeta, "get", new Object[] {refUnwrapped});
```
The method can be **invoked as static** method on the `Ref` type with first argument being the `Ref` object.

JaroslavTulach reviewed (approved):
The idea of using **static methods** on `Type` seems to be a great solution to all the problems we had so far. The PR is simple and unlocks a functionality that was clearly missing.

- cannot call `Ref.get`: https://github.com/enso-org/enso/pull/11577/files#r1848224116
- cannot call `File.path` on [EnsoFile](https://github.com/enso-org/enso/pull/11530/files#r1838708041)
- cannot call `to_text` on `Vector` - discovered while working on #11569

Can you simplify these three known workarounds in the production code as well? Then this PR will immediately be useful for Enso users!

Akirathan commented:
The approach to export methods as internal members of types and not on atoms/builtin objects was chosen because of https://github.com/enso-org/enso/issues/12043 - at the moment, it is not possible to implement `hasMembers` interop message on builtin objects.

Commit:
Simplify usages of Ref.get and Pth.root

Akirathan commented:
> Can you simplify these three known workarounds in the production code as well? Then this PR will immediately be useful for Enso users!

@JaroslavTulach Usages of `Ref.get` and `Path.root` simplified in 0c97057ccbe6a3b852644507bd38ddfe7770ee85.

I tried to simplify `Warning.to_text` as well in `WithWarnings` like so:
```diff
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/WithWarnings.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/WithWarnings.java
index 7b8dce1d6e..8acf02eedd 100644
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/WithWarnings.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/WithWarnings.java
@@ -8,7 +8,9 @@ import com.oracle.truffle.api.dsl.Cached.Shared;
 import com.oracle.truffle.api.interop.ArityException;
 import com.oracle.truffle.api.interop.InteropLibrary;
 import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.interop.UnknownIdentifierException;
 import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
 import com.oracle.truffle.api.library.CachedLibrary;
 import com.oracle.truffle.api.library.ExportLibrary;
 import com.oracle.truffle.api.library.ExportMessage;
@@ -201,6 +203,7 @@ public final class WithWarnings extends EnsoObject {
     for (var w : warns) {
       try {
         var wText = node.execute(toText, state, new Object[] {w.getValue()});
+        var myText = invokeToText(w, where);
         if (wText instanceof Text t) {
           if (prefix != null) {
             text = text.add(prefix);
@@ -214,6 +217,19 @@ public final class WithWarnings extends EnsoObject {
     return text;
   }
 
+  private static Object invokeToText(Warning warn, Node where) {
+    System.out.println("[mylog] invokeToText");
+    var interop = InteropLibrary.getUncached();
+    try {
+      var meta = interop.getMetaObject(warn);
+      var toTextRes = interop.invokeMember(meta, "to_text", warn);
+      return toTextRes;
+    } catch (UnsupportedMessageException | ArityException | UnknownIdentifierException |
+             UnsupportedTypeException e) {
+      throw EnsoContext.get(where).raiseAssertionPanic(where, e.getMessage(), e);
+    }
+  }
+
   @ExportMessage
   Object send(Message message, Object[] args, @CachedLibrary(limit = "3") ReflectionLibrary lib)
       throws Exception {
```
But that does not work, because `to_text` is defined on `Any`, and the [Type.methods](https://github.com/enso-org/enso/blob/83f4bcd330ce06294b0d0fb115756006656c9c0a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java#L505) method discovery does not include methods from Any, neither extension methods. I have created issue for it to track it down - https://github.com/enso-org/enso/issues/12045

Let's merge this PR with the current changes.

Commit:
Merge branch 'develop' into wip/akirathan/11589-builtins-expose-methods

# Conflicts:
#	distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso
